### PR TITLE
Remove the rest of libzerocash.

### LINKED ISSUE #510: Make it harder to misuse util.cpp in libzerocash
The API util.cpp presents is confusing. Some output sizes are specified via input sizes, and some of them resize for you and others don't. Decide what the "safest" API should be, noting that in C++ it's usually possible to make it cheap to return objects, even big vectors, because of move constructors and things like that. Then, make them all behave similarly.

From #493.

### PR DESCRIPTION
**Pull request by ebfull**
Closes #510.

### CODE REVIEW

**Comment by daira** [comment]
Reviewing.

File: src/gtest/test_libzcash_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#include <gtest/gtest.h>
+#include "zcash/util.h"
+
+TEST(libzcash_utils, convertBytesVectorToVector)
+{
+	std::vector<unsigned char> bytes = {0x00, 0x01, 0x03, 0x12, 0xFF};
+    std::vector<bool> expected_bits = {
```
Comment (daira) [inline src/gtest/test_libzcash_utils.cpp]: Is this mixing tabs and spaces?

Comment (ebfull) [inline src/gtest/test_libzcash_utils.cpp]: Doh, yes. This whole file probably is full of nasty tabs. I'll clean it up.

File: src/zcash/util.cpp [hunk 2]
```diff
@@ -10,3 +11,35 @@ std::vector<unsigned char> convertIntToVectorLE(const uint64_t val_int) {
 
     return bytes;
 }
+
+// Convert bytes into boolean vector.
+std::vector<bool> convertBytesVectorToVector(const std::vector<unsigned char>& bytes) {
+	std::vector<bool> ret;
+	ret.resize(bytes.size() * 8);
```
Comment (daira) [inline src/zcash/util.cpp]: Inconsistent indentation.

File: src/zcash/util.cpp [hunk 3]
```diff
@@ -10,3 +11,35 @@ std::vector<unsigned char> convertIntToVectorLE(const uint64_t val_int) {
 
     return bytes;
 }
+
+// Convert bytes into boolean vector.
```
Comment (daira) [inline src/zcash/util.cpp]: Add a comment that the bits in each byte are encoded from most-significant to least significant.

File: src/zcash/util.cpp [hunk 4]
```diff
@@ -10,3 +11,35 @@ std::vector<unsigned char> convertIntToVectorLE(const uint64_t val_int) {
...
+        c = bytes.at(i);
+        for (size_t j = 0; j < 8; j++) {
+            ret.at((i*8)+j) = (c >> (7-j)) & 1;
+        }
+    }
+
+    return ret;
+}
+
+// Convert boolean vector to integer
```
Comment (daira) [inline src/zcash/util.cpp]: Add a comment that the boolean vector is interpreted as little-endian.

File: src/zcash/util.cpp [hunk 5]
```diff
@@ -10,3 +11,35 @@ std::vector<unsigned char> convertIntToVectorLE(const uint64_t val_int) {
...
+uint64_t convertVectorToInt(const std::vector<bool>& v) {
+    if (v.size() > 64) {
+        throw std::length_error ("boolean vector can't be larger than 64 bits");
+    }
+
+    uint64_t result = 0;
+    for (size_t i=0; i<v.size();i++) {
+        if (v.at(i)) {
```
Comment (daira) [inline src/zcash/util.cpp]: This is not constant-time; is it used in any contexts where that would matter? Maybe just add a comment on #785 to check this.

**Comment by daira** [comment]
ACK modulo comments.

File: src/zcash/util.cpp (hunk 5 above)
Comment (ebfull) [inline src/zcash/util.cpp]: Done!

**Comment by ebfull** [comment]
Fixed @daira's nits. Up for review again!

**Comment by daira** [comment]
Sorry, my requested clarification for ``convertVectorToInt`` was wrong, it is actually big-endian. I'm still a bit confused because I didn't think we were still using big-endian integer encodings anywhere in the protocol. Where is this used?

**Comment by ebfull** [comment]
`convertVectorToInt` is only used for the weird libsnark API for merkle tree authentication paths, not for any of our encodings. I think that we can remove it entirely someday.

**Comment by ebfull** [comment]
Fixed, rebased.

**Comment by daira** [comment]
ACK.

**Comment by ebfull** [comment]
@zkbot r+

**Comment by ebfull** [comment]
Test suite was trying to run `utilTest` which no longer exists.

@zkbot r+

**Pull request merged**